File: web/schemas/support.py
from typing import List, Optional

from pydantic import BaseModel, Field

from web.schemas.types import UTCDatetime


class AttachmentResponse(BaseModel):
    id: int
    url: str
    content_type: Optional[str] = None
    file_size: Optional[int] = None

    model_config = {"from_attributes": True}


class SupportMessageResponse(BaseModel):
    id: int
    sender_type: str
    body: str
    created_at: UTCDatetime
    attachments: List[AttachmentResponse] = []

    model_config = {"from_attributes": True}


class SupportTicketListItem(BaseModel):
    id: int
    subject: str
    category: str
    status: str
    unread_by_user: bool
    last_message_at: UTCDatetime
    created_at: UTCDatetime

    model_config = {"from_attributes": True}


class SupportTicketListResponse(BaseModel):
    items: List[SupportTicketListItem]
    total: int
    page: int
    page_size: int


class SupportTicketDetailResponse(BaseModel):
    id: int
    subject: str
    category: str
    status: str
    unread_by_user: bool
    created_at: UTCDatetime
    last_message_at: UTCDatetime
    messages: List[SupportMessageResponse] = []

    model_config = {"from_attributes": True}


class CreateTicketRequest(BaseModel):
    subject: str = Field(min_length=1, max_length=200)
    category: str
    body: str = Field(min_length=1, max_length=4000)
    attachment_ids: List[int] = []


class CreateMessageRequest(BaseModel):
    body: str = Field(default="", max_length=4000)
    attachment_ids: List[int] = []


class UnreadCountResponse(BaseModel):
    count: int

File: frontend/src/api/support.ts
import { apiRequest, apiUpload, API_BASE } from './client'

export type TicketStatus = 'new' | 'in_progress' | 'closed'
export type TicketCategory = 'payment' | 'connection' | 'subscription' | 'other'

export interface Attachment {
  id: number
  url: string
  content_type: string | null
  file_size: number | null
}

export interface SupportMessage {
  id: number
  sender_type: 'user' | 'admin'
  body: string
  created_at: string
  attachments: Attachment[]
}

export interface SupportTicketListItem {
  id: number
  subject: string
  category: TicketCategory
  status: TicketStatus
  unread_by_user: boolean
  last_message_at: string
  created_at: string
}

export interface SupportTicketListResponse {
  items: SupportTicketListItem[]
  total: number
  page: number
  page_size: number
}

export interface SupportTicketDetail {
  id: number
  subject: string
  category: TicketCategory
  status: TicketStatus
  unread_by_user: boolean
  created_at: string
  last_message_at: string
  messages: SupportMessage[]
}

export interface CreateTicketRequest {
  subject: string
  category: TicketCategory
  body: string
  attachment_ids: number[]
}

export function listTickets(page = 0, pageSize = 50): Promise<SupportTicketListResponse> {
  return apiRequest<SupportTicketListResponse>(`/support/tickets?page=${page}&page_size=${pageSize}`)
}

export function getTicket(id: number): Promise<SupportTicketDetail> {
  return apiRequest<SupportTicketDetail>(`/support/tickets/${id}`)
}

export function createTicket(body: CreateTicketRequest): Promise<SupportTicketDetail> {
  return apiRequest<SupportTicketDetail>('/support/tickets', {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function addMessage(id: number, body: string, attachmentIds: number[]): Promise<SupportMessage> {
  return apiRequest<SupportMessage>(`/support/tickets/${id}/messages`, {
    method: 'POST',
    body: JSON.stringify({ body, attachment_ids: attachmentIds }),
  })
}

export function closeTicket(id: number): Promise<SupportTicketDetail> {
  return apiRequest<SupportTicketDetail>(`/support/tickets/${id}/close`, { method: 'POST' })
}

export function uploadSupportImage(file: File): Promise<Attachment> {
  const form = new FormData()
  form.append('file', file)
  return apiUpload<Attachment>('/support/attachments', form)
}

export function getSupportUnread(): Promise<{ count: number }> {
  return apiRequest<{ count: number }>('/support/unread')
}

export function requestSupportStreamTicket(): Promise<{ ticket: string }> {
  return apiRequest<{ ticket: string }>('/support/stream-ticket', { method: 'POST' })
}

export function buildSupportStreamUrl(ticket: string): string {
  return `${API_BASE}/support/stream?ticket=${encodeURIComponent(ticket)}`
}
